File: libs/service_layer/search/SearchEngine/src/main/kotlin/net/dankito/service/search/specific/TagsSearchResult.java
package net.dankito.service.search.specific;


import net.dankito.deepthought.model.Tag;

import java.util.ArrayList;
import java.util.List;


public class TagsSearchResult {

  protected String searchTerm;

  protected boolean hasExactMatch;

  protected Tag exactMatch;

  protected Tag singleMatch = null;

  protected List<Tag> allMatches;


  public TagsSearchResult(String searchTerm, List<Tag> allMatches) {
    this.searchTerm = searchTerm;
    this.allMatches = allMatches;

    findExactMatch(searchTerm, allMatches);
  }

  public TagsSearchResult(String searchTerm, List<Tag> allMatches, Tag exactMatch) {
    this.searchTerm = searchTerm;
    this.allMatches = allMatches;
    this.hasExactMatch = exactMatch != null;
    this.exactMatch = exactMatch;
  }


  protected void findExactMatch(String searchTerm, List<Tag> allMatches) {
    for(Tag match : allMatches) {
      if(searchTerm.equals(match.getName().toLowerCase())) {
        this.hasExactMatch = true;
        this.exactMatch = match;
        break;
      }
    }
  }


  public String getSearchTerm() {
    return searchTerm;
  }

  public boolean hasExactMatch() {
    return hasExactMatch;
  }

  public Tag getExactMatch() {
    return exactMatch;
  }

  public boolean hasSingleMatch() {
    return getAllMatchesCount() == 1;
  }

  public Tag getSingleMatch() {
    if(hasSingleMatch()) {
      if(singleMatch == null) {
        if(allMatches instanceof List)
          singleMatch = ((List<Tag>)allMatches).get(0);
        else
          singleMatch = new ArrayList<Tag>(allMatches).get(0);
      }

      return singleMatch;
    }

    return null;
  }

  public int getAllMatchesCount() {
    return allMatches.size();
  }

  public List<Tag> getAllMatches() {
    return allMatches;
  }


  @Override
  public String toString() {
    return searchTerm + " has " + getAllMatchesCount() + " matches, hasExactMatch = " + hasExactMatch();
  }

}

File: libs/service_layer/search/SearchEngine/src/main/kotlin/net/dankito/service/search/specific/TagsSearchResult.kt
package net.dankito.service.search.specific


import net.dankito.deepthought.model.Tag
import java.util.*


class TagsSearchResult {

    var searchTerm: String
        private set

    private var hasExactMatches: Boolean = false

    var exactMatches: List<Tag> = ArrayList()
        private set

    private var singleMatchProperty: Tag? = null

    var allMatches: List<Tag>
        private set


    constructor(searchTerm: String, allMatches: List<Tag>) {
        this.searchTerm = searchTerm
        this.allMatches = allMatches

        if(searchTerm.isNullOrBlank() == false) { // don't call findExactMatch() for an empty search term, would load all tags
            findExactMatch(searchTerm, allMatches)
        }
    }

    constructor(searchTerm: String, allMatches: List<Tag>, exactMatches: List<Tag>) {
        this.searchTerm = searchTerm
        this.allMatches = allMatches
        this.hasExactMatches = exactMatches.isNotEmpty()
        this.exactMatches = exactMatches
    }


    private fun findExactMatch(searchTerm: String, allMatches: List<Tag>) {
        for(match in allMatches) {
            if(match != null && searchTerm == match.name.toLowerCase()) {
                this.hasExactMatches = true
                (this.exactMatches as MutableList).add(match)
            }
        }
    }

    fun hasExactMatches(): Boolean {
        return hasExactMatches
    }

    fun hasSingleMatch(): Boolean {
        return allMatchesCount == 1
    }

    fun getSingleMatch(): Tag? {
        if (hasSingleMatch()) {
            if (singleMatchProperty == null) {
                if (allMatches is List<*>)
                    singleMatchProperty = allMatches[0]
                else
                    singleMatchProperty = ArrayList(allMatches)[0]
            }

            return singleMatchProperty
        }

        return null
    }

    val allMatchesCount: Int
        get() = allMatches.size


    override fun toString(): String {
        return searchTerm + " has " + allMatchesCount + " matches, hasExactMatches = " + hasExactMatches()
    }

}
